02-frequency_filter.py

```diff
@@ -40,32 +40,31 @@ def run_filter(subject, run=None, session=None):
     # room recording we wish to save.
     # The basenames of the empty-room recording output file does not contain
     # the "run" entity.
-    shared_basename_kws = dict(subject=subject,
-                               session=session,
-                               task=config.get_task(),
-                               acquisition=config.acq,
-                               processing=config.proc,
-                               recording=config.rec,
-                               space=config.space)
-
-    bids_basename = make_bids_basename(run=run, **shared_basename_kws)
-    bids_er_out_basename = make_bids_basename(**shared_basename_kws)
+    bids_basename = make_bids_basename(subject=subject,
+                                       run=run,
+                                       session=session,
+                                       task=config.get_task(),
+                                       acquisition=config.acq,
+                                       processing=config.proc,
+                                       recording=config.rec,
+                                       space=config.space,
+                                       prefix=deriv_path)
+
+    bids_er_out_basename = bids_basename.copy().update(run=None)
 
     # Prepare a name to save the data
     if config.use_maxwell_filter:
-        raw_fname_in = op.join(deriv_path, f'{bids_basename}_sss_raw.fif')
-        raw_er_fname_in = op.join(deriv_path,
-                                  f'{bids_basename}_emptyroom_sss_raw.fif')
+        raw_fname_in = bids_basename.copy().update(suffix='sss_raw.fif')
+        raw_er_fname_in = (bids_basename.copy()
+                           .update(suffix='emptyroom_sss_raw.fif'))
     else:
-        raw_fname_in = op.join(deriv_path, f'{bids_basename}_nosss_raw.fif')
-        raw_er_fname_in = op.join(
-            deriv_path,
-            f'{bids_basename}_emptyroom_nosss_raw.fif')
-
-    raw_fname_out = op.join(deriv_path, f'{bids_basename}_filt_raw.fif')
-    raw_er_fname_out = op.join(
-        deriv_path,
-        f'{bids_er_out_basename}_emptyroom_filt_raw.fif')
+        raw_fname_in = bids_basename.copy().update(suffix='nosss_raw.fif')
+        raw_er_fname_in = (bids_basename.copy()
+                           .update(suffix='emptyroom_nosss_raw.fif'))
+
+    raw_fname_out = bids_basename.copy().update(suffix='filt_raw.fif')
```

Comment:
this makes me wonder if `bids_basename.update` should always return a copy to avoid this awkward construction everywhere.

Reply:
actually never mind, I think this is fine.

Reply:
I was actually thinking the exact same thing, but wanted to bring this up at some later time ;) But since you're putting this forward:

I think it would be neater & cleaner from a user's perspective to be able to do sth like:

```python
bidspath_factory = BIDSPath(subject='foo', session='bar', is_factory=True)

fname_1 = bidspath_factory(suffix='a.fif')
fname_2 = bidspath_factory(suffix='b.fif')
fname_3 = bidspath_factory(subject='baz')
```

which would produce

```python
fname_1 == 'sub-foo_ses-bar_a.fif'
fname_2 == 'sub-foo_ses-bar_b.fif'
fname_3 == 'sub-baz_ses-bar'
```

Reply:
humm I am not too familiar with factory classes/methods. Why use that in place of a bool argument called `copy` that we use in MNE

Reply:
I discussed this with @agramfort and we'll leave it the way it is for now, i.e. `.copy().update(...)` :)